### PR TITLE
Remove Spec.DatasetName and Spec.Spec

### LINKED ISSUE #3101: Can we remove `Spec.Spec` ?
I'm not clear why it exists and it prevents people from using the parsed spec struct like a spec, because if they change any of the value of the struct, that field will be out of date.

It doesn't look like it is used for anything.

### PR DESCRIPTION
**Pull request by aboodman**
Fixes #3101

### CODE REVIEW

**Comment by aboodman** [comment]
jenkins: test this

**Comment by aboodman** [comment]
I thought I already landed this. Weird.

**Comment by aboodman** [comment]
@kalman ptal -- I have reworked this completely since last time.

File: go/spec/absolute_path.go [hunk 1]
```diff
@@ -16,15 +16,21 @@ import (
...
-// remainder of the path.
+// AbsolutePath describes the location of a Value within a Noms database.
+//
+// To locate a value relative to some other value, see Path. To locate a value
+// globally, see Spec.
+//
+// For more on paths, absolute paths, and specs, see:
+// https://github.com/attic-labs/noms/blob/master/doc/spelling.md.
 type AbsolutePath struct {
+	// The dataset this AbsolutePath is rooted at. Only one of Dataset and Hash should be set.
```
Comment (kalman) [inline go/spec/absolute_path.go]: "Dataset is the dataset..."

File: go/spec/absolute_path.go [hunk 2]
```diff
@@ -16,15 +16,21 @@ import (
...
 	Dataset string
-	Hash    hash.Hash
-	Path    types.Path
+	// The hash this AbsolutePath is rooted at. Only one of Dataset and Hash should be set.
+	Hash hash.Hash
```
Comment (kalman) [inline go/spec/absolute_path.go]: "Hash is the hash..."

File: go/spec/absolute_path.go [hunk 3]
```diff
@@ -16,15 +16,21 @@ import (
...
+	// The relative path from Dataset or Hash. This can be empty. In that case, the AbsolutePath
```
Comment (kalman) [inline go/spec/absolute_path.go]: "Path is the relative..."
Godoc should also be wrapped at 80 characters.

File: go/spec/spec.go [hunk 4]
```diff
@@ -91,21 +82,27 @@ func ForDataset(spec string) (Spec, error) {
...
 	if err != nil {
 		return Spec{}, err
 	}
 
-	sp.DatasetName = dsName
+	if path.Dataset == "" {
+		return Spec{}, errors.New("dataset name required for dataset spec")
+	}
+
+	if !sp.Path.Path.IsEmpty() {
```
Comment (kalman) [inline go/spec/spec.go]: Seems like this should be `!path.Path.IsEmpty()`. `sp.Path` hasn't been set yet. Is there no test that catches this?

File: go/spec/spec.go [hunk 5]
```diff
@@ -229,16 +241,11 @@ func (sp Spec) Pin() (Spec, bool) {
...
-	if sp.Path.Path != nil {
-		spec += sp.Path.Path.String()
-	}
-
-	pinned, err := ForPathOpts(spec, sp.Options)
-	d.PanicIfError(err)
-	*pinned.db = *sp.db
+	r := sp
+	r.Path.Hash = commit.Hash()
+	r.Path.Dataset = ""
```
Comment (kalman) [inline go/spec/spec.go]: This is so much nicer.

File: go/spec/spec_test.go [hunk 6]
```diff
@@ -44,8 +43,8 @@ func TestMemDatasetSpec(t *testing.T) {
 
 	assert.Equal("mem", spec.Protocol)
 	assert.Equal("", spec.DatabaseName)
-	assert.Equal("test", spec.DatasetName)
-	assert.True(spec.Path.IsEmpty())
+	assert.Equal("test", spec.Path.Dataset)
+	assert.True(spec.Path.Path.IsEmpty())
```
Comment (kalman) [inline go/spec/spec_test.go]: Perhaps:
```go
assert.Equal(AbsolutePath{
  Dataset: "test",
}, spec.Path)
```

File: go/spec/util.go [hunk 7]
```diff
@@ -5,19 +5,20 @@
...
+func CreateValueSpecString(protocol, db, path string) string {
+	p, err := NewAbsolutePath(path)
+	d.Chk.NoError(err)
+	return Spec{Protocol: protocol, DatabaseName: db, Path: p}.String()
 }
 
-func CreateHashSpecString(protocol, path string, h hash.Hash) string {
-	return fmt.Sprintf("%s:%s%s#%s", protocol, path, Separator, h.String())
+func CreateHashSpecString(protocol, db string, h hash.Hash) string {
+	return Spec{Protocol: protocol, DatabaseName: db, Path: AbsolutePath{Hash: h}}.String()
```
Comment (kalman) [inline go/spec/util.go]: This is so much nicer. (I'd break the objects over multiple lines, but whatever).

File: go/spec/spec_test.go (hunk 6 above)
Comment (kalman) [inline go/spec/spec_test.go]: There may be other tests you can improve by using this pattern. I'm not sure. There were surprisingly few changes in the test file actually.

File: go/spec/spec.go (hunk 4 above)
Comment (aboodman) [inline go/spec/spec.go]: Good catch. Added test.

File: go/spec/spec.go (hunk 5 above)
Comment (aboodman) [inline go/spec/spec.go]: \o/

File: go/spec/spec_test.go (hunk 6 above)
Comment (aboodman) [inline go/spec/spec_test.go]: Shrug.

**Pull request merged**